src/main/java/org/elasticsearch/index/IndexService.java

```
@@ -465,16 +467,23 @@ private void onShardClose(ShardLock lock, boolean ownsShard) {
     private class StoreCloseListener implements Store.OnClose {
         private final ShardId shardId;
         private final boolean ownsShard;
+        private final Closeable[] toClose;
 
-        public StoreCloseListener(ShardId shardId, boolean ownsShard) {
+        public StoreCloseListener(ShardId shardId, boolean ownsShard, Closeable... toClose) {
             this.shardId = shardId;
             this.ownsShard = ownsShard;
+            this.toClose = toClose;
         }
 
         @Override
         public void handle(ShardLock lock) {
-            assert lock.getShardId().equals(shardId) : "shard id mismatch, expected: " + shardId + " but got: " + lock.getShardId();
-            onShardClose(lock, ownsShard);
+            try {
+                assert lock.getShardId().equals(shardId) : "shard id mismatch, expected: " + shardId + " but got: " + lock.getShardId();
+                onShardClose(lock, ownsShard);
+            } finally {
+                IOUtils.closeWhileHandlingException(toClose);
```

Do we always want to ignore exceptions here?